src/frontend/src/optimizer/plan_node/logical_multi_join.rs

```diff
@@ -43,61 +58,189 @@ impl fmt::Display for LogicalMultiJoin {
-            conjunctions.extend(new_on.conjunctions);
-        } else {
-            inputs.push(right.clone());
+        let left = join.left();
+        let right = join.right();
+
+        let mut builder = Self::new(left);
+
+        let (r_output_indices, r_conjunctions, mut r_inputs, r_tot_input_col_num) =
+            Self::new(right).into_parts();
+
+        // the mapping from the right's column index to the current multi join's internal column
+        // index
+        let mut mapping = ColIndexMapping::with_shift_offset(
+            r_tot_input_col_num,
+            builder.tot_input_col_num as isize,
+        );
+        builder.inputs.append(&mut r_inputs);
+        builder.tot_input_col_num += r_tot_input_col_num;
+
+        builder.conjunctions.extend(
+            r_conjunctions
+                .into_iter()
+                .map(|expr| mapping.rewrite_expr(expr)),
+        );
+        builder
+            .conjunctions
+            .extend(join.on().conjunctions.iter().cloned());
+
+        builder
+            .output_indices
+            .extend(r_output_indices.into_iter().map(|idx| mapping.map(idx)));
+        builder.output_indices = join
+            .output_indices()
+            .iter()
+            .map(|idx| builder.output_indices[*idx])
+            .collect();
+        builder
+    }
+
+    fn with_filter(plan: PlanRef) -> LogicalMultiJoinBuilder {
+        let filter: &LogicalFilter = plan.as_logical_filter().unwrap();
+        let mut builder = Self::new(filter.input());
+        builder
+            .conjunctions
+            .extend(filter.predicate().conjunctions.clone());
+        builder
+    }
+
+    fn with_project(plan: PlanRef) -> LogicalMultiJoinBuilder {
+        let proj: &LogicalProject = plan.as_logical_project().unwrap();
+        let output_indices = match proj.try_as_projection() {
+            Some(output_indices) => output_indices,
+            None => return Self::with_input(plan),
+        };
+        let mut builder = Self::new(proj.input());
+        builder.output_indices = output_indices;
+        builder
+    }
+
+    fn with_input(input: PlanRef) -> LogicalMultiJoinBuilder {
+        LogicalMultiJoinBuilder {
+            output_indices: (0..input.schema().len()).collect_vec(),
+            conjunctions: vec![],
+            tot_input_col_num: input.schema().len(),
+            inputs: vec![input],
         }
+    }
 
-        Some(Self {
-            base: logical_join.base.clone(),
-            inputs,
-            on: Condition { conjunctions },
-        })
+    pub fn inputs(&self) -> &[PlanRef] {
+        self.inputs.as_ref()
     }
+}
+impl LogicalMultiJoin {
+    pub(crate) fn new(inputs: Vec<PlanRef>, on: Condition, output_indices: Vec<usize>) -> Self {
+        let input_schemas = inputs
+            .iter()
+            .map(|input| input.schema().clone())
+            .collect_vec();
+
+        let (inner_o2i_mapping, tot_col_num) = {
```

Comment:
Is it possible to re-use the logic in hash join? e.g. derive schema by left-deep two-way join instead of rewriting everything by ourselves.